src/noit_metric_director.c

```diff
@@ -154,76 +131,100 @@ noit_adjust_metric_interest(uuid_t id, const char *metric, short cnt) {
 }
 
 static void
-distribute_line(caql_cnt_t *interests,
-                const char *line, int line_len) {
+distribute_message_with_interests(caql_cnt_t *interests, metric_message_t *message) {
   int i;
-  for(i=0;i<nthreads;i++) {
+  for(i = 0; i < nthreads; i++) {
     if(interests[i] > 0) {
-      char *copy = mtev__strndup(line, line_len);
-      ck_fifo_spsc_t *fifo = (ck_fifo_spsc_t *)thread_queues[i];
+      ck_fifo_spsc_t *fifo = (ck_fifo_spsc_t *) thread_queues[i];
       ck_fifo_spsc_entry_t *fifo_entry;
       fifo_entry = ck_fifo_spsc_recycle(fifo);
       if(!fifo_entry) fifo_entry = malloc(sizeof(ck_fifo_spsc_entry_t));
       ck_fifo_spsc_enqueue_lock(fifo);
```

Comment:
This lock needs to move above the `recycle` line.  `recycle` is not MP safe so the lock should protect it.